include/diff.hpp

```diff
@@ -1,147 +1,21 @@
-            throw std::exception();
-        }
-    }
-
-    char operator[] (size_t i) const {
-        return m_str[m_view_index_begin + i];
-    }
-
-    size_t index_begin() const {
-        return m_view_index_begin;
-    }
-    
-    size_t index_end() const {
-        return m_view_index_end;
-    }
-
-    const char* begin() const noexcept {
-        return &m_str[m_view_index_begin];
-    }
-    
-    const char* end() const noexcept {
-        return &m_str[m_view_index_end];
-    }
-
-    size_t size() const {
-        return m_view_index_end - m_view_index_begin;
-    }
 
-    template <class t_os>
-    friend t_os& operator<< (t_os& os, const sequence_view& seq) {
-        for (size_t i = seq.m_view_index_begin; i < seq.m_view_index_end; i++) {
-            os << seq.m_str[i];
-        }
-
-        return os;
-    }
-};
-
-enum class PatchOperation {
-    Addition,
-    Deletion,
-    Undefined
-};
-
-class Patch {
-    PatchOperation m_op;
-    size_t m_begin;
-    std::string m_str;
-
-    Patch(PatchOperation op, size_t begin, std::string&& str) :
-        m_op(op),
-        m_begin(begin),
-        m_str(str)
-    { }
-
-    Patch(PatchOperation op, size_t begin, const std::string& str) :
-        m_op(op),
-        m_begin(begin),
-        m_str(str)
-    { }
-  public:
-
-    static Patch make_addition(const sequence_view& seq) {
-        auto str = std::string(std::cbegin(seq), seq.size());
-        return Patch(PatchOperation::Addition, seq.index_begin(), std::move(str));
-    }
+#include "patch.hpp"
+#include "sequence_view.hpp"
 
-    static Patch make_deletion(const sequence_view& seq) {
-        auto str = std::string(std::cbegin(seq), seq.size());
-        return Patch(PatchOperation::Deletion, seq.index_begin(), std::move(str));
-    }
-
-    PatchOperation GetOperation() {
-        return m_op;
-    }
-
-    template <class t_os>
-    friend t_os& operator<< (t_os& os, const Patch& p) {
-        os << "Patch ";
-
-        if (p.m_op == PatchOperation::Addition) {
-            os << "Addition";
-        } else if (p.m_op == PatchOperation::Deletion) {
-            os << "Deletion";
-        } else {
-            os << "Undefined";
-        }
-
-        os << ": [" << p.m_begin << ':' << p.m_begin + p.m_str.size() << "] - " << p.m_str;
-
-        return os;
-    }
-};
-
-class Diff {
-    std::vector<Patch> _diff(const sequence_view& lhs_seq, const sequence_view& rhs_seq) const {
-        int32_t lhs_size = lhs_seq.size();
-        int32_t rhs_size = rhs_seq.size();
-        int32_t max_len = lhs_size + rhs_size;
-        int32_t x_values_len = 2 * std::min(lhs_size, rhs_size) + 2;
+namespace diff {
+    template <class t_sequenced_type>
+    class _diff {
+        const sequence_view<t_sequenced_type>& lhs_seq;
+        const sequence_view<t_sequenced_type>& rhs_seq;
+        std::vector<patch>& patches;
+        int32_t lhs_size;
+        int32_t rhs_size;
+        int32_t max_len;
```

Comment:
Length should always be unsigned, `uint32_t` is the better option here. Even better is to use `size_t`